### ISSUE TITLE
Spells

### ISSUE THREAD

**Issue by descawed** [descawed/tesconvert#6]
Need to convert over the player's known spells and known magic effects. For spells that contain effects that can't be transferred over (like Levitate), do we cut the whole spell or just that effect (assuming there are other, convertible effects)? I'm leaning towards the whole spell, because the effects that don't exist (Levitate, Almsivi Intervention, Divine Intervention, Mark, Recall) are typically the primary component of a spell, and the spell is likely to be fundamentally altered without them. But we should also allow mapping these effects to effects added by mods.

**Comment by descawed** [comment]
We may have to do this before #8 because active magic effects have a reference to the spell they belong to. This means we need to not only convert the player's own spells, but also any spells currently effecting the player.

**Comment by descawed** [comment]
Things to consider when deciding how to convert spells:

- Morrowind spells have a range of possible magnitudes; Oblivion skills have a fixed magnitude. How do we choose a fixed magnitude to convert to? Could base it on the --combine flag like skills. [UESP's article on enchanting](https://en.uesp.net/wiki/Morrowind:Enchant#Constant_Effect_Tips) claims that Luck affects the probability of a particular magnitude being chosen, which could be an additional complicating factor. However, OpenMW doesn't seem to do that in CastSpell::inflict. Is that behavior specific to enchantments or is one of the two sources wrong?
- The cost for auto-calc spells will need to be recalculated. Refer to [this article](https://cs.elderscrolls.com/index.php?title=Category:Spell_Cost) on the Oblivion CS wiki.
- It's been a while since I played Oblivion, but I recall Oblivion spellmaking limiting spell power much more strictly than Morrowind. For instance, [this article](https://cs.elderscrolls.com/index.php?title=FMagicLevelMagnitudeMult) mentions that magnitude is capped at 100, and level-based effects are additionally capped based on fMagicLevelMagnitudeMult. Do we need to cap certain values when we convert or is it fine (i.e. won't break anything) to port overpowered Morrowind spells unaltered?
- Oblivion spells need to be grouped into tiers based on [these game settings](https://cs.elderscrolls.com/index.php?title=FMagicSpellLevelXXMin).
- Should we map vanilla Morrowind spells to vanilla Oblivion spells? The "equivalent" spells are usually not exactly the same. For instance, Morrowind's "absorb agility" spell absorbs 5-20 Agility for 30 seconds, whereas Oblivion's "StandardAbsorbAttribAgility1Novice" absorbs 5 Agility for 20 seconds. The alternative would be to convert every spell as a custom spell at the risk of having some duplicates. Honestly, I think it's better to convert everything; I can't think of any "unique" spells where you wouldn't want a duplicate. The only thing that might get annoying is I think there are one or two quests that require you to learn certain spells, and "equivalent" converted spells might not satisfy the requirement. That's a minor issue, though; we can address that at a later date.

**Comment by descawed** [comment]
Spell conversion has been implemented. I decided not to toss the entire spell if only some effects can't be converted; a spell will be converted if any of its effects can be converted.

One question I've left unanswered is what to do with an edge case for the Calm, Demoralize, and Rally effects. In Morrowind, each of these are two separate effects, one for creatures and one for humanoids. In Oblivion, there's only a single effect that affects both types of actors. For now, I'm not doing any special handling of this. If a Morrowind spell contains, e.g., both a Rally Creature and Rally Humanoid effect, the Oblivion spell will have two Rally effects. This is probably not the desired behavior, though; maybe we should have a check for this. That's not a pressing issue, however, so I'm going to close this issue for now.
